File: hacker-rank-interview-preparation-kit/Ruby/arrays/minimum_swaps_2.rb
=begin
Tae H0 K1m
  8. Minimu Swaps 2 (medium)

  You are given an unordered array consisting of consecutive integers; [1, 2, 3,
  ..., n] without any duplicates. You are allowed to swap any two elements.
  You need to find the minimum number of swaps required to sort the array in
  ascending order.

   It must return an integer representing the minimum number of swaps to sort
   the array.
=end

# 1. iterate from 1 to n
# 2. find i's index and swap with correct value
# 3. count!

def minimumSwaps(arr)
  swaps = 0
  i = 1
  arr.each do |element|
    j = arr.index(i) # i's index
    unless arr[j] == element
      arr[i-1] = arr[j]
      arr[j] = element
      swaps += 1
    end
    i += 1
  end
  return swaps
end


puts minimumSwaps([4,3,1,2])
# => 3

puts minimumSwaps([7,1,3,2,4,5,6])
# => 5

puts minimumSwaps([2,3,4,1,5])
# => 3

puts minimumSwaps([1,3,5,2,4,6,7])
# => 3

File: hacker-rank-interview-preparation-kit/Python/Arrays/minimum_swaps_2.py
"""
    8. Minimum Swaps 2
    Tae H0 K1m

    problem: Return the minimum number of swaps to sort the given array.

    input: the unsorted array
    output: the minimum number of swaps to sort the array

    approach:
        1. iterate through array
        2. if the current value is not correct position,
        3. find the correct position and
        4. swap the two values, and increment the number of swaps

        example:

        2 3 4 1 5
        1 2 3 4 5

        1 <--> 4
        1 3 4 2 5

        2 <--> 4
        1 2 4 3 5

        3 <--> 4
        1 2 3 4 5
"""
def find(arr, pos):
    for i in range(len(arr)):
        if pos == arr[i]:
            return i
    return -1

def swap(arr, i, index):
    temp = arr[i]
    arr[i] = arr[index]
    arr[index] = temp
    return arr

def minimumSwaps(arr):
    result = 0
    for i in range(len(arr)):
        pos = i + 1
        val = arr[i]
        if val != pos:
            index = find(arr, pos)
            arr = swap(arr, i, index)
            result += 1
    return result

print(minimumSwaps([2,3,4,1,5]))
# => 3

print(minimumSwaps([1,3,5,2,4,6,7]))
# => 3
